cmd/app/createca.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package app
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math"
+	"math/big"
+	"os"
+	"time"
+
+	"github.com/ThalesIgnite/crypto11"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"github.com/sigstore/fulcio/pkg/log"
+)
+
+// createcaCmd represents the createca command
+var createcaCmd = &cobra.Command{
+	Use:   "createca",
+	Short: "Create a root CA in a pkcs11 device",
+	Long: `Create an x509 root CA within a pkcs11 device using values
+such as organization, country etc. This can then be used as the root
+certificate authority for an instance of sigstore fulcio`,
+	Run: func(cmd *cobra.Command, args []string) {
+		log.Logger.Info("binding to PKCS11 HSM")
+		p11Ctx, err := crypto11.ConfigureFromFile("config/crypto11.conf")
+		if err != nil {
+			log.Logger.Fatal(err)
+		}
+		defer p11Ctx.Close()
+
+		rootID := []byte(viper.GetString("hsm-caroot-id"))
+
+		// Check if CA already exists (or a cert within the provided ID)
+		findCA, err := p11Ctx.FindCertificate(rootID, nil, nil)
+		if err != nil {
+			panic(err)
```

Comment:
why log.Fatal() some places but panic() in others?

Reply:
Good point, I read up more on both approaches and went with Log.Fatal, if these should be panic I can swap approach.